From 341f9637658a0039fcc55ed2597eaafcf68450b9 Mon Sep 17 00:00:00 2001
From: Saurabh Sengar <saurabh.singh@xilinx.com>
Date: Tue, 6 Feb 2018 13:49:46 +0530
Subject: [PATCH 0251/1852] drm: xlnx: adding cursor width / height support

commit 32eda2976ffd349510191cb1240af6017f9dff9d from
https://github.com/Xilinx/linux-xlnx.git

Adding cursor width / height support to xlnx crtc.
Any crtc driver with cursor width / height requirement will register
these callbacks

Signed-off-by: Saurabh Sengar <saurabh.singh@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_crtc.c | 29 +++++++++++++++++++++++++++++
 drivers/gpu/drm/xlnx/xlnx_crtc.h |  6 ++++++
 drivers/gpu/drm/xlnx/xlnx_drv.c  |  4 ++++
 3 files changed, 39 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_crtc.c b/drivers/gpu/drm/xlnx/xlnx_crtc.c
index de83905206f7..44abb5b549a7 100644
--- a/drivers/gpu/drm/xlnx/xlnx_crtc.c
+++ b/drivers/gpu/drm/xlnx/xlnx_crtc.c
@@ -128,6 +128,35 @@ uint32_t xlnx_crtc_helper_get_format(struct xlnx_crtc_helper *helper)
 	return format;
 }
 
+uint32_t xlnx_crtc_helper_get_cursor_width(struct xlnx_crtc_helper *helper)
+{
+	struct xlnx_crtc *crtc;
+	int width = XLNX_CRTC_MAX_HEIGHT_WIDTH, tmp;
+
+	list_for_each_entry(crtc, &helper->xlnx_crtcs, list) {
+		if (crtc->get_cursor_width) {
+			tmp = crtc->get_cursor_width(crtc);
+			width = min(width, tmp);
+		}
+	}
+
+	return width;
+}
+
+uint32_t xlnx_crtc_helper_get_cursor_height(struct xlnx_crtc_helper *helper)
+{
+	struct xlnx_crtc *crtc;
+	int height = XLNX_CRTC_MAX_HEIGHT_WIDTH, tmp;
+
+	list_for_each_entry(crtc, &helper->xlnx_crtcs, list) {
+		if (crtc->get_cursor_height) {
+			tmp = crtc->get_cursor_height(crtc);
+			height = min(height, tmp);
+		}
+	}
+
+	return height;
+}
 struct xlnx_crtc_helper *xlnx_crtc_helper_init(struct drm_device *drm)
 {
 	struct xlnx_crtc_helper *helper;
diff --git a/drivers/gpu/drm/xlnx/xlnx_crtc.h b/drivers/gpu/drm/xlnx/xlnx_crtc.h
index bff5e3d94957..b3c7056fd179 100644
--- a/drivers/gpu/drm/xlnx/xlnx_crtc.h
+++ b/drivers/gpu/drm/xlnx/xlnx_crtc.h
@@ -28,6 +28,8 @@
  * @get_max_width: Get the maximum supported width
  * @get_max_height: Get the maximum supported height
  * @get_format: Get the current format of CRTC device
+ * @get_cursor_width: Get the cursor width
+ * @get_cursor_height: Get the cursor height
  */
 struct xlnx_crtc {
 	struct drm_crtc crtc;
@@ -37,6 +39,8 @@ struct xlnx_crtc {
 	int (*get_max_width)(struct xlnx_crtc *crtc);
 	int (*get_max_height)(struct xlnx_crtc *crtc);
 	uint32_t (*get_format)(struct xlnx_crtc *crtc);
+	uint32_t (*get_cursor_width)(struct xlnx_crtc *crtc);
+	uint32_t (*get_cursor_height)(struct xlnx_crtc *crtc);
 };
 
 /*
@@ -50,6 +54,8 @@ u64 xlnx_crtc_helper_get_dma_mask(struct xlnx_crtc_helper *helper);
 int xlnx_crtc_helper_get_max_width(struct xlnx_crtc_helper *helper);
 int xlnx_crtc_helper_get_max_height(struct xlnx_crtc_helper *helper);
 uint32_t xlnx_crtc_helper_get_format(struct xlnx_crtc_helper *helper);
+uint32_t xlnx_crtc_helper_get_cursor_width(struct xlnx_crtc_helper *helper);
+uint32_t xlnx_crtc_helper_get_cursor_height(struct xlnx_crtc_helper *helper);
 
 struct xlnx_crtc_helper *xlnx_crtc_helper_init(struct drm_device *drm);
 void xlnx_crtc_helper_fini(struct drm_device *drm,
diff --git a/drivers/gpu/drm/xlnx/xlnx_drv.c b/drivers/gpu/drm/xlnx/xlnx_drv.c
index ef7821a59627..971b0a9286de 100644
--- a/drivers/gpu/drm/xlnx/xlnx_drv.c
+++ b/drivers/gpu/drm/xlnx/xlnx_drv.c
@@ -129,6 +129,10 @@ static void xlnx_mode_config_init(struct drm_device *drm)
 	drm->mode_config.min_height = 0;
 	drm->mode_config.max_width = xlnx_crtc_helper_get_max_width(crtc);
 	drm->mode_config.max_height = xlnx_crtc_helper_get_max_height(crtc);
+	drm->mode_config.cursor_width =
+		xlnx_crtc_helper_get_cursor_width(crtc);
+	drm->mode_config.cursor_height =
+		xlnx_crtc_helper_get_cursor_height(crtc);
 }
 
 static void xlnx_lastclose(struct drm_device *drm)
-- 
2.31.1

